README.md
# SSL Certificate Automation Script (BASH)

To quickly and easily create a self-signed SSL certificate for Apache and Nginx web servers I wrote a little script in BASH.

![Screenshot of the Script in Action](screenshot.png)

This repository contains a BASH script that automates the creation of self-signed SSL certificates for Apache and Nginx web servers. The goal is to simplify and streamline the process of SSL certificate generation, allowing for secure network information transfers with minimal effort.

SSL certificates are required to ensure the secure transfer of information in the network. Self-signed SSL certificates, in the realm of cryptography and computer security, are identity certificates that are signed by the same entity they certify. Essentially, if you create the SSL certificate for your own domain or IP address, it will be self-signed. These certificates are most suited for internal use such as on an intranet.


## Usage

1. Before running the script, ensure it has the required executable permissions. This only needs to be done once, prior to the first execution of the script:

```bash
chmod +x ssl_crt_creater.sh
```

2. Run the `ssl_crt_creater.sh` script:
```bash
./ssl_crt_creater.sh
```

After the SSL certificate is created, you can then bind it to your server.


## Dependencies and Instructions

This script requires the `dialog` and `openssl` packages. The `dialog` package is used for menu rendering, while `openssl` is used for certificate creation. If these packages are not already installed, the script will prompt you to do so.

The commands used for certificate and key creation are as follows:

For NginX:

```bash
openssl req -new -x509 -days 365 -nodes -out /etc/nginx/ssl/$__servername.crt -keyout /etc/nginx/ssl/$__servername.key
```

For Apache:

```bash
openssl req -new -x509 -days 365 -nodes -out /etc/apache2/ssl/$__servername.crt -keyout /etc/apache2/ssl/$__servername.key
```

Here is a brief description of the command arguments:

- `req` - Request to create a new certificate.
- `-new` - Creates a Certificate Signing Request (CSR).
- `-x509` - Creates a self-signed certificate instead of a CSR.
- `-days 365` - Sets the certificate validity period to 365 days (1 year).
- `-nodes` - Prevents the private key from being encrypted.
- `-out` - Specifies the location to store the certificate.
- `-keyout` - Specifies the location to store the private key.

Upon execution, the script will generate a new certificate and a private RSA key with a length of 2048 bits. These will be placed in the respective working directories for Apache (`/etc/apache2/ssl/`) and Nginx (`/etc/nginx/ssl/`) with file permissions set to `600` for security reasons.

The generated private key and certificate can be found at:

- For Apache: `/etc/apache2/ssl/*.crt` and `/etc/apache2/ssl/*.key`
- For NginX: `/etc/nginx/ssl/*.crt` and `/etc/nginx/ssl/*.key`


## Contributing

Welcome and thanks! I appreciate you taking the initiative to contribute to this project.

Contributing isn’t limited to just code. I encourage you to contribute in the way that best fits your abilities, by writing tutorials, making translation to your native language, giving a demo at your local meetup, helping other users with their support questions, or revising  the documentation for this project.

Please take a moment to read the guidelines in the [CONTRIBUTING.md](CONTRIBUTING.md). Following them helps to communicate that you respect the time of the other contributors to the project. In turn, they’ll do their best to reciprocate that respect when working with you, across timezones and around the world.


## Security Vulnerabilities

If you discover a security vulnerability within this script, kindly email me directly. All security vulnerabilities will be promptly addressed.

## License

This script is an open-source software licensed under the [MIT License](LICENSE.md) and is distributed free of charge.

Commercial licensing (for projects that cannot use an open-source license) is available upon request.


## Author

Arthur Garegnyan

- Email: arthurgareginyan@gmail.com
- GitHub: [https://github.com/ArthurGareginyan/](https://github.com/ArthurGareginyan/)
- Website: [http://www.arthurgareginyan.com](http://www.arthurgareginyan.com)
- Donate: [http://www.arthurgareginyan.com/donate.html](http://www.arthurgareginyan.com/donate.html)


CHANGELOG.md

# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.4.0] - 2023-07-22

- **Support for multiple package managers**: Updated the `checkNeededPackages` function to use the available package manager (`apt-get`, `dnf`, `yum`, or `zypper`) on the system. If no known package manager is found, the script will display a message and exit.
- **User-specified certificate validity**: Added functions `setValidityPeriod` and `checkValidityPeriod` to allow the user to specify the validity period of the certificate (in days). This value is used when generating the certificate.
- **Code comments for improved readability**: Introduced extensive inline comments to the script code. These comments aim to enhance readability and maintainability of the code by explaining the functionality of each segment of the script. This enhancement aids both developers and users in understanding and potentially adjusting the script according to their needs.


## [1.3.0] - 2023-07-21

- **Removed `sudo` usage within script**: The script already requires root privileges to run, and there's a function to check whether the script is run as root. So, `sudo` is no longer used when installing packages.
- **Introduced variables for directory paths**: Directory paths like `/etc/nginx/ssl` and `/etc/apache2/ssl` are now represented by variables. This makes it easier to maintain and modify the script in the future.
- **Merged certificate generation functions**: The parts of the script that generate SSL certificates for Nginx and Apache were very similar and have been merged into a single function. This function takes arguments to reduce repetition and improve maintainability.
- **Removed hard-coded server types**: The server type (NginX or Apache) is now passed as an argument to the certificate generation function. This improves the modularity of the script.

## [1.2.0] - 2020-05-01

## [1.1.0] - 2015-11-04

## [1.0.0] - 2015-05-25
- Initial version of the SSL Certificate Creator script.


ssl_crt_creater.sh

#!/bin/bash

#=============================================================#
# Name:         SSL Certificate Creater                       #
# Description:  Create a self-signed SSL Certificates         #
#               for Apache and Nginx web-servers.             #
# Version:      1.4                                           #
# Released:     25.05.2015                                    #
# Updated:      21.07.2023                                    #
# Author:       Arthur Gareginyan                             #
# Author URI:   http://www.arthurgareginyan.com               #
# Email:        arthurgareginyan@gmail.com                    #
# License:      MIT License                                   #
# License URI:  http://opensource.org/licenses/MIT            #
#=============================================================#

#                  USAGE:
#        chmod +x ./ssl_crt_creater.sh
#        sudo ./ssl_crt_creater.sh


################## DECLARE FUNCTIONS ######################

# Function to check if the script is run as root
checkRoot() {
    if [ $(id -u) -ne 0 ];
        then
            printf "Script must be run as root. Try 'sudo ./ssl_certificate_creater.sh'\n"
            exit 1
    fi
}

# Function to check if the necessary packages are installed and install them if necessary
checkNeededPackages() {
    # List of necessary packages
    lst="dialog openssl"
    
    # Loop over the list of packages
    for items in $lst
    do
        # If a package is not installed
        type -P $items &>/dev/null || {
            echo -en "\n Package \"$items\" is not installed!"
            echo -en "\n Install now? [yes]/[no]: "
            read ops
            case $ops in
                YES|yes|Y|y)
                    # If 'apt-get' is available, use it to install the package
                    if [ -n "$(command -v apt-get)" ]; then
                        apt-get install $items
                    # Else, if 'dnf' is available, use it to install the package
                    elif [ -n "$(command -v dnf)" ]; then
                        dnf install $items
                    # Else, if 'yum' is available, use it to install the package
                    elif [ -n "$(command -v yum)" ]; then
                        yum install $items
                    # Else, if 'zypper' is available, use it to install the package
                    elif [ -n "$(command -v zypper)" ]; then
                        zypper install $items
                    else
                        # If no known package manager is found, print an error message and exit the script
                        echo "No known package manager found"
                        exit 1
                    fi ;;
                *)
                    # If the user chooses not to install the necessary packages, print a message and exit the script
                    echo -e "\n Exiting..."
                    exit 1 ;;
            esac
        }
    done
}

# Function to get the server name from the user
setServerName() {
    # Display a dialog box and store the user's input in the 'choices' variable
    cmd=(dialog --backtitle "mycyberuniverse.com - Create SSL Certificate for NGinX/Apache" \
        --inputbox "\n Please enter the URL of your website." 22 76 $__servername)
    choices=$("${cmd[@]}" 2>&1 >/dev/tty)
    
    # If the user entered a server name, store it in the '__servername' variable
    if [ "$choices" != "" ];
        then
            __servername=$choices
        else
            break
    fi
}

# Function to get the certificate validity period from the user
setValidityPeriod() {
    # Display a dialog box and store the user's input in the 'choices' variable
    cmd=(dialog --backtitle "mycyberuniverse.com - Create SSL Certificate for NGinX/Apache" \
        --inputbox "\n Please enter the validity period of the certificate (in days)." 22 76 $validity_period)
    choices=$("${cmd[@]}" 2>&1 >/dev/tty)
    
    # If the user entered a validity period, store it in the 'validity_period' variable
    if [ "$choices" != "" ];
        then
            validity_period=$choices
        else
            break
    fi
}

# Function to check if the server name has been set
checkServerName() {
    if [ "$__servername" = "" ];
        then
            setServerName
    fi
}

# Function to check if the certificate validity period has been set
checkValidityPeriod() {
    if [ "$validity_period" = "" ];
        then
            setValidityPeriod
    fi
}

# Function to generate a certificate for a specific server type and certificate directory
generateCertificate() {
    # Store the function arguments in variables
    server_type=$1
    cert_dir=$2

    # Display a message to the user
    dialog --backtitle "mycyberuniverse.com - Create SSL Certificate for NGinX/Apache" \
           --title "Create SSL Certificate for $server_type" \
           --msgbox "\n We are now going to create a self-signed certificate. While you could simply press ENTER when you are asked for country name etc. or enter whatever you want, it might be beneficial to have the web servers host name in the common name field of the certificate." 20 60

    # Create the certificate directory if it doesn't exist
    mkdir -p $cert_dir

    # Generate the certificate and store it in the certificate directory
    openssl req -new -x509 -days $validity_period -nodes -out $cert_dir/$__servername.crt -keyout $cert_dir/$__servername.key

    # Change the permissions of the key file
    chmod 600 $cert_dir/$__servername.key

    # Display a message to the user
    dialog --backtitle "mycyberuniverse.com - Create SSL Certificate for NGinX/Apache" \
           --title "Create SSL Certificate for $server_type" \
           --msgbox "\n Done! Your certificates are available at $cert_dir/$__servername.crt & $cert_dir/$__servername.key" 20 60
}


######################## GO ###############################

# Check if the script is run as root
checkRoot

# Check if the necessary packages are installed
checkNeededPackages

# Main loop
while true;
do
    # Display a menu to the user
    cmd=(dialog --backtitle "mycyberuniverse.com - Create SSL Certificate for NGinX/Apache" \
                --title "Create SSL Certificate for NGinX/Apache" \
                --menu "You MUST set the server URL (e.g., myaddress.dyndns.org) and certificate validity period (in days) before starting create certificate. Choose task:" 20 60 15)
    options=(1 "Set server URL ($__servername)"
             2 "Set certificate validity period ($validity_period)"
             3 "Generate new SSL certificate for NGiNX"
             4 "Generate new SSL certificate for Apache"
             5 "Exit")
    choice=$("${cmd[@]}" "${options[@]}" 2>&1 >/dev/tty)
    if [ "$choice" != "" ];
        then
            case $choice in
                1) setServerName ;;
                2) setValidityPeriod ;;
                3) checkServerName
                   checkValidityPeriod
                   generateCertificate "NGinX" "/etc/nginx/ssl" ;;
                4) checkServerName
                   checkValidityPeriod
                   generateCertificate "Apache" "/etc/apache2/ssl" ;;
                5) clear
                   exit 0 ;;
            esac
        else
            break
    fi
done
clear

exit 0
